docs/fsharp-cheatsheet.md

```diff
@@ -540,17 +538,156 @@ Another way of implementing interfaces is to use *object expressions*.
-    
-    let fizzBuzz = function 
-        | DivisibleBy 3 & DivisibleBy 5 -> "FizzBuzz" 
-        | DivisibleBy 3 -> "Fizz" 
-        | DivisibleBy 5 -> "Buzz" 
+
+    let fizzBuzz = function
+        | DivisibleBy 3 & DivisibleBy 5 -> "FizzBuzz"
+        | DivisibleBy 3 -> "Fizz"
+        | DivisibleBy 5 -> "Buzz"
         | i -> string i
 
 *Partial active patterns* share the syntax of parameterized patterns but their active recognizers accept only one argument.
 
+<a name="CodeOrganization"></a>Code Organization
+---------
+
+<a name="CodeOrganization_Modules"></a>
+### Modules
+Modules are key building blocks for grouping related code; they can contain `types`, `let` bindings, or (nested) sub `module`s.
+Identifiers within modules can be referenced using dot notation, or you can bring them into scope via the `open` keyword. Illustrative-only example:
+
+    module Game =
+        let mutable basePoints = 1
+        type Player = { id: int; score: int }
+        let playerScored player = { player with score = player.score + basePoints }
+
+    let player: Game.Player = { id = 1; score = 0 }
+    let player' = Game.playerScored player  // score = 1
+
+    open Game
+    basePoints <- 2
+    let player'' = playerScored player'  // player''.score = 3
+
+If you have only one module in a file, the `module` name can be declared at the top of the file, with all further declarations
+being module elements (and non indentation required)
+
+    module Functions  // notice there is no '=' when at the top of a file
+
+    let addToFive num = 5 + num
+    let subtractFive num = num - 5
+
+### Namespaces
+Namespaces are simply dotted names that prefix other program elements to allow for further hierarchical organization.
+All `type` and `module` elements that follow a `namespace` declaration will require an [`open`](#CodeOrganization_OpenAndAutoOpen) or to be dotted-into to access.
+If a new `namespace` is specified, all elements following will be part of the new namespace.
+
+    namespace MyNamespace
+
+    namespace MyNamespace.SubNamespace
+
+They can also be specified in a file-level [`module`](#CodeOrganization_Modules) definition, but no further `namespace` declarations may follow.
+
+    module MyNamespace.SubNamespace.Functions
+
+<a name="CodeOrganization_OpenAndAutoOpen"></a>
+### Open and AutoOpen
+
+The `open` keyword can be used with `module`, `namespace`, and `type`.
+
+    open System.Diagnostics  // open namespace
+    let stopwatch = Stopwatch.StartNew()
+    ---
+    module MyModule =
+        type DU1 = A | B | C
+        type DU2 = D | E | F
+
+    open type MyModule.DU1
+    let du1 = A
+    let duNotDefined = D  // 'D' not defined
+    open MyModule
+    let du2 = D
+
+Available to `module` elements, is the `AutoOpen` attribute. This alleviates the need for an `open`; *however* this should be used cautiously,
+as all following declarations will be immediately brought into the global namespace and cause conflicts.
```

Comment:
'following' is a bit ambiguous; maybe all elements/contents
'the global namespace' -> 'the scope whenever the parent is `open`ed, which can cause ambiguities'
(basically its not a conflict, which would be clear and unambiguous - its shadowing, and things become messy as the winning name is defined by the open order)

Perhaps this section should nod to the fact that every open can shadow names, but that in general you want to avoid that coming into play except in very select circumstances.

(People have all sorts of things like topological grouping of opens i.e. system stuff, then domain stuff etc but IME having no shadowing and doing one group sorted alphabetically is probably the best overall)

Other things here are that opens should generally specify the full namespace, i.e. no `open System` then `open Globalization`